In [1]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

dokumen = [
    "Cinta Luar Biasa",
    "Hati-Hati di Jalan",
    "Sial",
    "Tak Ingin Usai",
    "Bertaut",
    "Runtuh",
    "Komang",
    "Pesan Terakhir",
    "Melukis Senja",
    "Satu Bulan"
]

queries = {
    "cinta hati": {1, 2},
    "jalan terakhir": {2, 8},
    "runtuh usai": {4, 6}
}

vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(dokumen)


def precision_at_k(hasil, relevan, k):
    return sum(1 for d in hasil[:k] if d in relevan) / k


def recall(hasil, relevan):
    if not relevan:
        return 0.0
    return sum(1 for d in hasil if d in relevan) / len(relevan)


def f1(p, r):
    if p + r == 0:
        return 0.0
    return 2 * p * r / (p + r)


def average_precision(hasil, relevan):
    if not relevan:
        return 0.0

    hit = 0
    total = 0.0

    for i, d in enumerate(hasil, start=1):
        if d in relevan:
            hit += 1
            total += hit / i

    return total / len(relevan)


nilai_ap = []

for query, relevan in queries.items():

    print("\n" + "=" * 70)
    print("QUERY:", query)
    print("=" * 70)

    query_vector = vectorizer.transform([query])
    similarity = cosine_similarity(query_vector, tfidf_matrix).flatten()

    ranking = similarity.argsort()[::-1]
    hasil = [index + 1 for index in ranking]

    print("\n5 DOKUMEN TERATAS")

    for peringkat, index in enumerate(ranking[:5], start=1):
        print(f"\nPeringkat {peringkat}")
        print("ID Dokumen :", index + 1)
        print("Judul      :", dokumen[index])
        print("Skor       :", f"{similarity[index]:.4f}")

    p = precision_at_k(hasil, relevan, 5)
    r = recall(hasil, relevan)
    f = f1(p, r)
    ap = average_precision(hasil, relevan)

    nilai_ap.append(ap)

    print("\nGROUND TRUTH :", relevan)
    print("Precision@5  :", f"{p:.4f}")
    print("Recall       :", f"{r:.4f}")
    print("F1-Score     :", f"{f:.4f}")
    print("AP           :", f"{ap:.4f}")


MAP = sum(nilai_ap) / len(nilai_ap)

print("\n" + "=" * 70)
print("HASIL AKHIR")
print("=" * 70)
print("MAP :", f"{MAP:.4f}")


QUERY: cinta hati

5 DOKUMEN TERATAS

Peringkat 1
ID Dokumen : 2
Judul      : Hati-Hati di Jalan
Skor       : 0.5774

Peringkat 2
ID Dokumen : 1
Judul      : Cinta Luar Biasa
Skor       : 0.4082

Peringkat 3
ID Dokumen : 9
Judul      : Melukis Senja
Skor       : 0.0000

Peringkat 4
ID Dokumen : 10
Judul      : Satu Bulan
Skor       : 0.0000

Peringkat 5
ID Dokumen : 8
Judul      : Pesan Terakhir
Skor       : 0.0000

GROUND TRUTH : {1, 2}
Precision@5  : 0.4000
Recall       : 1.0000
F1-Score     : 0.5714
AP           : 1.0000

QUERY: jalan terakhir

5 DOKUMEN TERATAS

Peringkat 1
ID Dokumen : 8
Judul      : Pesan Terakhir
Skor       : 0.5000

Peringkat 2
ID Dokumen : 2
Judul      : Hati-Hati di Jalan
Skor       : 0.2887

Peringkat 3
ID Dokumen : 9
Judul      : Melukis Senja
Skor       : 0.0000

Peringkat 4
ID Dokumen : 10
Judul      : Satu Bulan
Skor       : 0.0000

Peringkat 5
ID Dokumen : 6
Judul      : Runtuh
Skor       : 0.0000

GROUND TRUTH : {8, 2}
Precision@5  : 0.4000
Recall    

======================================================================
QUERY: cinta hati
======================================================================

5 DOKUMEN TERATAS

Peringkat 1
ID Dokumen : 2
Judul      : Hati-Hati di Jalan
Skor       : 0.5774

Peringkat 2
ID Dokumen : 1
Judul      : Cinta Luar Biasa
Skor       : 0.4082

Peringkat 3
ID Dokumen : 9
Judul      : Melukis Senja
Skor       : 0.0000

Peringkat 4
ID Dokumen : 10
Judul      : Satu Bulan
...
======================================================================
HASIL AKHIR
======================================================================
MAP : 1.0000
Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...